File: fixedWindow.go
package main

import (
	"strconv"
	"time"
)

func RateLimitUsingFixedWindow(userID string, intervalInSeconds int64, maximumRequests int64) bool {
	// userID can be apikey, location, ip
	currentWindow := strconv.FormatInt(time.Now().Unix()/intervalInSeconds, 10)
	key := userID + ":" + currentWindow // user userID + current time window
	// get current window count
	value, _ := redisClient.Get(ctx, key).Result()
	requestCount, _ := strconv.ParseInt(value, 10, 64)
	if requestCount >= maximumRequests {
		// drop request
		return false
	}

	// increment request count by 1
	redisClient.Incr(ctx, key) // if the key is not available, value is initialised to 0 and incremented to 1

	// handle request
	return true
	// delete all expired keys at regular intervals
}

File: fixedWindow.ts
import { Redis } from "ioredis";

const redisClient = new Redis();

async function rateLimitUsingFixedWindow(
  userID: string,
  intervalInSeconds: number,
  maximumRequests: number
): Promise<boolean> {
  const currentWindow = Math.floor(
    Date.now() / 1000 / intervalInSeconds
  ).toString();
  const key = `${userID}:${currentWindow}`;

  // Get current window count
  const value = await redisClient.get(key);
  const requestCount = value ? parseInt(value, 10) : 0;

  if (requestCount >= maximumRequests) {
    // Drop request
    return false;
  }

  // Increment request count by 1
  await redisClient.incr(key);

  // Handle request
  return true;

  // Note: Redis TTL for key expiration can be handled elsewhere depending on use case.
}
